File: src/mysqlcookbook/recipes/transfer/cvt_date.py
#!/usr/bin/python3
# cvt_date.py: Read input data, look for values that match a date
# format, convert them to another format. Also converts 2-digit
# years to 4-digit years on request. By default, assumes a
# transition point of 70, but this can be changed.

# Assumes tab-delimited, linefeed-terminated input lines.

# To add another format:
# - Modify argument processing code
# - Add a case to the input value test
# - Add a case to the output date formatting test

# Shortcoming: Doesn't accommodate trailing times (for date/time values)

import sys
import re
import getopt
import fileinput
import warnings

def check_date_format(format, type):
  if re.match('^iso', format, re.IGNORECASE):
    return "iso"
  if re.match('^us', format, re.IGNORECASE):
    return "us"
  if re.match('^br', format, re.IGNORECASE):
    return "british"
  sys.exit("Invalid %s date format: %s" % (type, format))

# Determine whether a year is a leap year.  Requires 4-digit
# year value (e.g., "1993", not "93").

def is_leap_year(year):
  return ((year % 4 == 0) and ((year % 100 != 0) or (year % 400 == 0) ) )

# Given year and month, return the number of days in the month.  Year
# value must be in 4-digit form or the leap-year test won't be correct.

def days_in_month(year, month):
  day_tbl = [31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31] 
  days = day_tbl[month - 1]

  if month == 2 and is_leap_year(year):
    days += 1
  return days

# Given an array of integers representing year, month, and day of a date,
# determine whether the date is valid.

def is_valid_date(year, month, day):
  # year must be nonnegative, month and day must be positive
  if year < 0 or month < 0 or day < 1:
    return 0
  # check maximum limits on individual parts
  if year > 9999 or month > 12 or day > days_in_month(year, month):
    return 0
  return 1

prog = "cvt_date.py"
usage = f'''
Usage: {prog} [ options ] [ data_file ]

Options:
--help
    Print this message
--iformat=format_name, --if=format_name
    Set date format for input (iso, us, british; default is iso)
--oformat=format_name, --of=format_name
    Set date format for output (iso, us, british; default is iso)
--warn
    Warn about bad dates
--add-century
    Convert 2-digit years to 4 digits
--transition
    Within-century transition point for 2-digit to 4-digit conversion
    (implies --add-century); default is 70
--columns=column_list
    Convert only specific columns; column_list is a comma-separated list
    of column numbers (first column is numbered 1)

Other arguments:
data_file
    Name of data file to convert
'''

iformat = "iso"
oformat = "iso"
transition = 70
warn = False
add_century = False
filter_columns = False
columns = {}
yearfmt = "%02d"

long_options = ["help", "if=", "iformat=", "of=", "oformat=", "warn", "add-century", "transition="]

try:
  opts, args = getopt.getopt(sys.argv[1:], '', long_options)
except getopt.GetoptError as err:
  print(usage)
  sys.exit(2)

for o, a in opts:
  if o == "--help":
    print(usage)
    sys.exit(0)
  if o in("--if", "--iformat"):
    iformat = check_date_format(a, "input")
  if o in("--of", "--oformat"):
    oformat = check_date_format(a, "output")
  if o == "--warn":
    warn = True
  if o == "--add-century":
    add_century = True
    yearfmt = "%04d"
  if o == "--transition":
    transition = a
  if o == "--columns":
    cols = a.split(",")
    for c in cols:
      if re.match('^\d+$', c):
        if int(c) <= 0:
          sys.exit("Column specifier %s is not a positive integer" % (c))
        columns[int(c)] = 1
      else:
        m = re.match('^(\d+)-(\d+)$', c)
        if not m:
          sys.exit("%s is not a valid column specifier" % (c))
        (begin, end) = (int(m.group(1)), int(m.group(2)))
        if begin < 0 or end < 0 or begin > end:
          sys.exit("%s is not a valid column specifier" % (c))
        while begin <= end:
          columns[begin] = 1
          begin += 1
    filter_columns = True

# 2-digit years don't provide accurate leap year testing
if warn and not add_century:
  warnings.warn("NOTE: --warn without --add-century may produce spurious warnings")

line_num = 0
for line in fileinput.input(args):
  line_num += 1
  values = line.rstrip().split("\t", 10000)  # split, preserving all fields
  for i in range(0, len(values)):
    # don't process the value if a column list was
    # specified but this column isn't one of them
    if filter_columns and i not in columns:
      continue

    val = values[i]
    if iformat == "iso":
      m = re.match('^(\d{2}|\d{4})\D(\d{1,2})\D(\d{1,2})$', val)
      if m:
        (y, m, d) = (int(m.group(1)), int(m.group(2)), int(m.group(3)))
      else:
        continue
    elif iformat == "us":
      m = re.match('^(\d{1,2})\D(\d{1,2})\D(\d{2}|\d{4})$', val)
      if m:
        (m, d, y) = (int(m.group(1)), int(m.group(2)), int(m.group(3)))
      else:
        continue
    elif iformat == "british":
       m = re.match('^(\d{1,2})\D(\d{1,2})\D(\d{2}|\d{4})$', val)
       if m:
         (d, m, y) = (int(m.group(1)), int(m.group(2)), int(m.group(3)))
       else:
         continue
    else:  # not a recognizable date; go to next value
      continue

    # convert 2-digit year to 4 digits on request
    if add_century and y < 100:
      y += 1900 if y >= transition else 2000

    if warn and not is_valid_date(y, m, d):
      warnings.warn("line %d: %s invalid?" % (line_num, val))

    if oformat == "iso":
      val = f"{yearfmt}-%02d-%02d" % (y, m, d)
    elif oformat == "us":
      val = f"%02d-%02d-{yearfmt}" % (m, d, y)
    elif oformat == "british":
      val = f"%02d-%02d-{yearfmt}" % (d, m, y)
    else:
      # should never happen!
      sys.exit("Logic error: unknown output format %s" % (oformat));
    values[i] = val  # replace original value with new one
  print ("\t".join(values))

File: src/mysqlcookbook/recipes/transfer/cvt_date.pl
#!/usr/bin/perl
# cvt_date.pl: Read input data, look for values that match a date
# format, convert them to another format. Also converts 2-digit
# years to 4-digit years on request. By default, assumes a
# transition point of 70, but this can be changed.

# Assumes tab-delimited, linefeed-terminated input lines.

# To add another format:
# - Modify argument processing code
# - Add a case to the input value test
# - Add a case to the output date formatting test

# Shortcoming: Doesn't accommodate trailing times (for date/time values)

use strict;
use warnings;
use Getopt::Long;
$Getopt::Long::ignorecase = 0; # options are case sensitive
$Getopt::Long::bundling = 1;   # permit short options to be bundled

sub check_date_format
{
my ($format, $type) = @_;

  return "iso" if $format =~ /^iso/i;
  return "us" if $format =~ /^us/i;
  return "british" if $format =~ /^br/i;
  die "Invalid $type date format: $format\n";
}

# Determine whether a year is a leap year.  Requires 4-digit
# year value (e.g., "1993", not "93").

sub is_leap_year
{
my $year = shift;

  return ($year % 4 == 0) && ((($year % 100) != 0) || ($year % 400) == 0);
}

# Given year and month, return the number of days in the month.  Year
# value must be in 4-digit form or the leap-year test won't be correct.

sub days_in_month
{
my ($year, $month) = @_;
my @day_tbl = (31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31);
my $days = $day_tbl[$month-1];

  # add a day for Feb of leap years
  $days++ if $month == 2 && is_leap_year ($year);
  return $days;
}

# Given an array of integers representing year, month, and day of a date,
# determine whether the date is valid.

sub is_valid_date
{
my ($year, $month, $day) = @_;

  # year must be nonnegative, month and day must be positive
  return 0 if $year < 0 || $month < 1 || $day < 1;
  # check maximum limits on individual parts
  return 0 if $year > 9999;
  return 0 if $month > 12;
  return 0 if $day > days_in_month ($year, $month);
  return 1;
}

my $prog = "cvt_date.pl";
my $usage = <<EOF;
Usage: $prog [ options ] [ data_file ]

Options:
--help
    Print this message
--iformat=format_name, --if=format_name
    Set date format for input (iso, us, british; default is iso)
--oformat=format_name, --of=format_name
    Set date format for output (iso, us, british; default is iso)
--warn
    Warn about bad dates
--add-century
    Convert 2-digit years to 4 digits
--transition
    Within-century transition point for 2-digit to 4-digit conversion
    (implies --add-century); default is 70
--columns=column_list
    Convert only specific columns; column_list is a comma-separated list
    of column numbers (first column is numbered 1)

Other arguments:
data_file
    Name of data file to convert
EOF

my $help;

my $iformat;     # input format
my $oformat;     # output format
my $warn;        # issue warnings about bad date
my $add_century; # convert 2-digit years to 4 digits
my $transition;  # transition at which 2-digit years are assumed to be 19XX
my $columns;     # check only these columns

GetOptions (
  # =s means a string value is required after the option
  "help"          => \$help,        # print help message
  "if=s"          => \$iformat,     # set input format only
  "iformat=s"     => \$iformat,     # ditto
  "of=s"          => \$oformat,     # set output format only
  "oformat=s"     => \$oformat,     # ditto
  "warn"          => \$warn,        # warn about bad dates
  "add-century"   => \$add_century, # convert 2-digit years to 4 digits
  "transition=n"  => \$transition,  # within-century transition point
                    # (implies --add-century)
  "columns=s"     => \$columns      # check only specific columns
) or die "$usage\n";

die "$usage\n" if defined ($help);

$iformat = "iso" unless defined ($iformat);     # set default format
$oformat = "iso" unless defined ($oformat);
$iformat = check_date_format ($iformat, "input");
$oformat = check_date_format ($oformat, "output");

$transition = 70 unless defined ($transition);    # set default transition

# 2-digit years don't provide accurate leap year testing
warn "NOTE: --warn without --add-century may produce spurious warnings\n"
                      if $warn && !$add_century;

my %columns;
if (defined ($columns))
{
  # make sure column specifiers are positive integers and hash them
  # (also convert from 1-based to 0-based values)
  foreach my $i (split (/,/, $columns))
  {
    if ($i =~ /^\d+$/)        # single column number
    {
      die "Column specifier $i is not a positive integer\n"
          unless $i > 0;
      $columns{$i-1} = 1;
    }
    elsif ($i =~ /^(\d+)-(\d+)$/) # column range m-n
    {
      my ($begin, $end) = ($1, $2);
      die "$i is not a valid column specifier\n"
          unless $begin > 0 && $end > 0 && $begin <= $end;
      while ($begin <= $end)
      {
        $columns{$begin-1} = 1;
        ++$begin;
      }
    }
    else
    {
      die "$i is not a valid column specifier\n";
    }
  }
}
$columns = defined ($columns);  # use as boolean flag from this point

# format for sprintf'ing year values; width depends on whether
# or not 2-digit to 4-digit conversion was requested

my $yearfmt = ($add_century ? "%04d" : "%02d");

my ($y, $m, $d);

while (<>)
{
  chomp;
  my @val = split (/\t/, $_, 10000);  # split, preserving all fields
  for my $i (0 .. @val - 1)
  {
    # don't process the value if a column list was
    # specified but this column isn't one of them
    next if $columns && !exists ($columns{$i});

    my $val = $val[$i];
    if ($iformat eq "iso"
      && $val =~ /^(\d{2}|\d{4})\D(\d{1,2})\D(\d{1,2})$/)
    {
      ($y, $m, $d) = ($1, $2, $3);
    }
    elsif ($iformat eq "us"
      && $val =~ /^(\d{1,2})\D(\d{1,2})\D(\d{2}|\d{4})$/)
    {
      ($m, $d, $y) = ($1, $2, $3);
    }
    elsif ($iformat eq "british"
      && $val =~ /^(\d{1,2})\D(\d{1,2})\D(\d{2}|\d{4})$/)
    {
      ($d, $m, $y) = ($1, $2, $3);
    }
    else  # not a recognizable date; go to next value
    {
      next;
    }

    # convert 2-digit year to 4 digits on request
    $y += ($y >= $transition ? 1900 : 2000) if $add_century && $y < 100;

    warn "line $.: $val invalid?\n"
      if $warn && !is_valid_date ($y, $m, $d);

    if ($oformat eq "iso")
    {
      $val = sprintf ("$yearfmt-%02d-%02d", $y, $m, $d);
    }
    elsif ($oformat eq "us")
    {
      $val = sprintf ("%02d-%02d-$yearfmt", $m, $d, $y);
    }
    elsif ($oformat eq "british")
    {
      $val = sprintf ("%02d-%02d-$yearfmt", $d, $m, $y);
    }
    else
    {
      # should never happen!
      die "Logic error: unknown output format $oformat\n";
    }
    $val[$i] = $val;  # replace original value with new one
  }
  print join ("\t", @val) . "\n";
}
